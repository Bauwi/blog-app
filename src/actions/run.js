import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  console.log(post);
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  console.log(DBid);
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      const run = [];
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      dispatch(setRunPosts(run));
    });
};
