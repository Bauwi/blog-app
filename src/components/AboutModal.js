import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Modal } from 'antd';

export class AboutModal extends Component {
  state = {
    visible: false,
    confirmLoading: false
  };
  showModal = () => {
    this.setState({
      visible: true
    });
  };

  handleCancel = () => {
    this.setState({
      visible: false
    });
  };
  render() {
    const { visible, confirmLoading } = this.state;
    return (
      <div className="login__modal">
        <div onClick={this.showModal}>
          <i className="fa fa-question" /> About
        </div>
        <Modal
          title="About"
          visible={visible}
          confirmLoading={confirmLoading}
          onCancel={this.handleCancel}
          footer={[]}
        >
          <div className="footer__modal">
            <p>This website's single purpose is to talk about art.</p>
            <br />
            <p>This website is a learning project.</p>
            <br />
            <p>
              All rights of articles are reserved to their author, who are credited at the beginning
              of each article.
            </p>
            <br />
            <p>
              If you are an author and want one of the article to be removed, please contact us via
              the contact page.
            </p>
          </div>
        </Modal>
      </div>
    );
  }
}

export default AboutModal;
