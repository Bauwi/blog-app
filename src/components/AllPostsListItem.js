import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

const AllPostsListItem = ({
  id, title, createdAt, shortBody, keywords, authorId
}) => (
  <div>
    <Link to={`${authorId}/read/${id}`}>
      <header>
        <h3>{title}</h3>
        <h5>Published {moment(createdAt).format('MMM Do, YYYY')}</h5>
      </header>
      <p>{shortBody}</p>
      <p>Keywords: {keywords}</p>
    </Link>
  </div>
);

export default AllPostsListItem;
