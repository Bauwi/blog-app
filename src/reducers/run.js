const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
