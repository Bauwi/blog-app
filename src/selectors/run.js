const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
