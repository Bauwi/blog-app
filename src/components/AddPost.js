import React, { Component } from 'react';
import PostForm from './PostForm';
import { connect } from 'react-redux';
import { startAddPost, startAddStar } from '../actions/posts';

export class AddPost extends Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h1>New Post</h1>
        <PostForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post)),
  startAddStar: id => dispatch(startAddStar(id))
});

export default connect(undefined, mapDispatchToProps)(AddPost);
