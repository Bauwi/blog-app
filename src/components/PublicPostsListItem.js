import React, { Component } from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { message } from 'antd';

import KeywordsList from './KeywordsList';

import { startAddPostToRun, startRemovePostToRun } from '../actions/run';

const success = () => {
  message.success('Post added to your run.');
};

const error = () => {
  message.error('Your run can not contain more than 20 reads.');
};

export class PublicPostsListItem extends Component {
  handleAddPostToRun = () => {
    if (this.props.isInRun) {
      this.props.startRemovePostToRun(this.props.post.id, this.props.DBid);
    } else if (this.props.run.length < 20) {
      this.props.startAddPostToRun({ content: this.props.post, state: 'unread' });
      this.props.success();
    } else {
      this.props.error();
    }
  };

  render() {
    const { id, title, createdAt, author, keywords, authorId, miniCover, stars } = this.props.post;
    const buttonClassName = this.props.isInRun
      ? 'button--pin button--pin--selected'
      : 'button--pin';
    return (
      <div className="homelist-item">
        <Link to={`/${authorId}/read/${id}`} className="no-decoration">
          <div className="homelist-item__content">
            <header>
              <img
                src={miniCover}
                className="list-item__img homelist-item__content__cover"
                alt="post cover"
              />
            </header>
            <section className="homelist-item__content__header">
              <div>
                <i className="fa fa-star" /> {stars}
              </div>
              <p className="homelist-item__content__date">
                {moment(createdAt).format('MMM Do, YYYY')}
              </p>
            </section>
            <div className="homelist-item__content__infos">
              <h3 className="homelist-item__content__title">{title}</h3>

              <div className="homelist-item__content__subtitle">
                <p className="homelist-item__content__author">{author}</p>
              </div>
            </div>
          </div>
        </Link>

        <footer className="homelist-item__footer">
          <KeywordsList keywords={keywords} />
          <div className="">
            <div className="button--pin__container">
              <p className="homelist-item__footer__index">{this.props.index}</p>
              <button className={buttonClassName} onClick={this.handleAddPostToRun}>
                <i className="fa fa-book" />
              </button>
            </div>
          </div>
        </footer>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPostToRun: post => dispatch(startAddPostToRun(post)),
  startRemovePostToRun: (id, DBid) => dispatch(startRemovePostToRun(id, DBid)),
  success: () => success(),
  error: () => error()
});

const mapStateToProps = (state, props) => {
  const isInRun = state.run.posts
    ? !!state.run.posts.find(post => props.post.id === post.content.id)
    : undefined;
  const DBid = isInRun
    ? state.run.posts.find(post => props.post.id === post.content.id).DBid
    : undefined;
  return {
    run: state.run.posts || 0,
    isInRun,
    index: isInRun ? state.run.posts.findIndex(post => props.post.id === post.content.id) + 1 : '',
    DBid
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(PublicPostsListItem);
