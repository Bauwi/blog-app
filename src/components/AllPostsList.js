import React, { Component } from 'react';
import { connect } from 'react-redux';

import { startSetPostsSample } from '../actions/posts';
import selectPosts from '../selectors/posts';

import AllPostsListItem from './AllPostsListItem';
import PostListFilters from './PostListFilters';

export class AllPostsList extends Component {
  componentWillMount() {
    this.props.startSetPostsSample(20);
  }

  renderListItem = () => {
    return this.props.posts.map(post => {
      return <AllPostsListItem {...post} />;
    });
  };

  render() {
    return (
      <div>
        <PostListFilters />
        <p>ALLpost List Here</p>
        {this.renderListItem()}
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetPostsSample: sampleSize => dispatch(startSetPostsSample(sampleSize))
});

const mapStateToProps = state => ({
  posts: selectPosts(state.posts, state.filters)
});

export default connect(mapStateToProps, mapDispatchToProps)(AllPostsList);
