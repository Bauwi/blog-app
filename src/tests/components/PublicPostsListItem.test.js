import React from 'react';
import { shallow } from 'enzyme';
import { PublicPostsListItem } from '../../components/PublicPostsListItem';
import run from '../fixtures/run';

test('should render PublicPostsListItem component properly', () => {
  const wrapper = shallow(<PublicPostsListItem
    post={run.posts[1].content}
    run={run.posts}
    startAddPostToRun={() => {}}
    startRemovePostToRun={() => {}}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startAddPostToRun on click', () => {
  const startAddPostToRun = jest.fn();
  const success = jest.fn();
  const wrapper = shallow(<PublicPostsListItem
    post={run.posts[1].content}
    run={[]}
    startAddPostToRun={startAddPostToRun}
    startRemovePostToRun={() => {}}
    success={success}
    error={() => {}}
  />);
  wrapper
    .find('button')
    .at(0)
    .simulate('click');
  expect(success).toHaveBeenCalled();
  expect(startAddPostToRun).toHaveBeenLastCalledWith({
    content: run.posts[1].content,
    state: 'unread'
  });
});

test('should call startRemovePostToRun on click', () => {
  const startRemovePostToRun = jest.fn();
  const wrapper = shallow(<PublicPostsListItem
    post={run.posts[1].content}
    run={run.posts}
    startAddPostToRun={() => {}}
    startRemovePostToRun={startRemovePostToRun}
    success={() => {}}
    error={() => {}}
    isInRun
    DBid="DBid"
  />);
  wrapper
    .find('button')
    .at(0)
    .simulate('click');
  expect(startRemovePostToRun).toHaveBeenLastCalledWith(run.posts[1].content.id, 'DBid');
});

test('should call error if run length is higher than 20', () => {
  const runTwenty = [];
  for (let i = 0; i <= 20; i += 1) {
    runTwenty.push({});
  }
  const error = jest.fn();
  const wrapper = shallow(<PublicPostsListItem
    post={run.posts[1].content}
    run={runTwenty}
    startAddPostToRun={() => {}}
    startRemovePostToRun={() => {}}
    success={() => {}}
    error={error}
  />);
  wrapper
    .find('button')
    .at(0)
    .simulate('click');
  expect(error).toHaveBeenCalled();
});
