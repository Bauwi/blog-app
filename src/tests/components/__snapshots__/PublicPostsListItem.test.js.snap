// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PublicPostsListItem component properly 1`] = `
<div
  className="homelist-item"
>
  <div>
    <div
      className="homelist-item__content"
    >
      <header>
        <img
          alt="post cover"
          className="list-item__img homelist-item__content__cover"
          src="https://www.w3schools.com/w3css/img_fjords.jpg"
        />
      </header>
      <div
        className="homelist-item__content__container"
      >
        <Link
          className="no-decoration"
          replace={false}
          to="/author2Id/read/2"
        >
          <div
            className="homelist-item__content__infos"
          >
            <section
              className="homelist-item__content__header"
            >
              <div>
                <i
                  className="fa fa-star"
                />
                 
                25
              </div>
              <p
                className="homelist-item__content__date"
              >
                Jan 1st, 1970
              </p>
            </section>
            <h3
              className="homelist-item__content__title"
            >
              Post 2 in Run
            </h3>
            <div
              className="homelist-item__content__subtitle"
            >
              <p
                className="homelist-item__content__author"
              >
                author2
              </p>
            </div>
          </div>
        </Link>
        <footer
          className="homelist-item__footer"
        >
          <KeywordsList
            keywords="cat, dog, fish"
          />
          <div
            className="homelist-item__footer__bottom"
          >
            <div>
              <p
                className="category border-cat"
              >
                cat
              </p>
            </div>
            <div
              className="button--pin__container"
            >
              <p
                className="homelist-item__footer__index"
              />
              <button
                className="button--pin"
                onClick={[Function]}
              >
                <i
                  className="fa fa-book"
                />
              </button>
            </div>
          </div>
        </footer>
      </div>
    </div>
  </div>
</div>
`;
